[USH] Case Search validation conditions

esoergel opened the pull request:
## Product Description
https://dimagi-dev.atlassian.net/browse/USH-2175
https://docs.google.com/document/d/1wF9SFfX7-R86cPRgabrqbTU9BaGdp3IlAYlTjInWXTU/edit
This adds the ability to configure validation conditions for case search input.  Sure feels an awful lot like we're rebuilding xforms.
![image](https://user-images.githubusercontent.com/2367539/177627182-b45e3cb0-1d97-4393-8d9d-90d4a7ae8363.png)

## Technical Summary
see commit messages for details

## Feature Flag
USH_CASE_CLAIM_UPDATES: USH Specific toggle to support several different case search/claim workflows in web apps

## Safety Assurance

### Safety story
Automated tests included for the suite changes, plus manual testing of the UI portion

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan
I believe this can be merged independently, then QA'd when the webapps and formplayer portions are complete.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

esoergel committed:
Add validation condition to model

esoergel committed:
Add validation conditions to suite xml

esoergel committed:
Make validation messages translatable

esoergel committed:
Update suite.xml format to agree with mobile

esoergel committed:
Make bucket for 2.53 USH case claim features

Thus implementing Simon's sarcastic comment from here:
https://github.com/dimagi/commcare-hq/pull/31568/files#r863234756

esoergel committed:
Build out minimally functional UI for validation

esoergel committed:
Update translations properly

This adds multiple language support for validation messages, refactors
it for labels, and fixes it for hints.  Previously, changing a hint in
one language would delete that hint for all other languages.

esoergel committed:
Only show validation message if xpath is set

esoergel commented on corehq/apps/app_manager/views/modules.py:
```diff
@@ -1009,22 +1009,19 @@ def _update_search_properties(module, search_properties, lang='en'):
-            label.update({lang: prop['label']})
-        else:
-            label = {lang: prop['label']}
-        hint = {lang: prop['hint']}
```
This was ignoring other languages and resetting the config to just include the current language

orangejenny replied:
nice fix

orangejenny commented on corehq/apps/app_manager/suite_xml/xml_models.py:
```diff
@@ -503,7 +503,7 @@ class Assertion(XmlObject):
 class Validation(XmlObject):
     ROOT_NAME = 'validation'
-    xpath = XPathField('@xpath')
+    test = XPathField('@test')
```
Two attributes, which differ by one letter, seems error-prone, but I defer to you & Shubham on final XML spec. (Possible alternative: `function` instead of `test`?)

esoergel replied:
No strong feelings here - I'd originally gone with `xpath`, but `Assertion` (just above this) also does `test` and `text`, so this seemed reasonable too.  @shubham1g5 any thoughts?

orangejenny replied:
ah, I didn't remember `Assertion` already does this, that lessens my complaint

shubham1g5 replied:
Yeah we used `test` to make them consistent with assertions.

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/details/case_claim.js:
```diff
@@ -346,6 +353,8 @@ hqDefine("app_manager/js/details/case_claim", function () {
                     exclude: searchProperties[i].exclude,
                     required: searchProperties[i].required,
+                    validationXPath: validation ? validation.xpath : '',
+                    validationMessage: validation ? validation.message[lang] : '',
```
Just pattern matching, it's odd that validationMessage references the language, but hint and label, both translatable strings, don't. Would it be doable to handle all of these the same way?

esoergel replied:
They do reference `lang`, it's just higher up in the function - I wouldn't object to moving those in-line if that's what you're thinking.
https://github.com/dimagi/commcare-hq/blob/fd574f67ddb3061685de878cd33c02b59c8dbf69/corehq/apps/app_manager/static/app_manager/js/details/case_claim.js#L321-L322

orangejenny replied:
aha, got it. I do tend to update this file in a very mechanical, copy-paste kind of way each time we add a new case search feature, but I don't think it's a big deal either way.

orangejenny commented on corehq/apps/app_manager/templates/app_manager/partials/modules/case_search_property.html:
```diff
@@ -185,6 +185,22 @@ <h4 class="modal-title">
+                            {% trans "Validation Condition" %}
+                          </label>
+                          <div class="col-xs-12 col-sm-9">
+                            <input class="form-control" type="text" data-bind="value: validationXPath" placeholder="true()"/>
```
UX: This should be a textarea, too, most xpath inputs get long.

orangejenny commented on corehq/apps/app_manager/app_strings.py:
```diff
@@ -185,6 +185,7 @@ def maybe_add_index(text):
             for prop in module.search_config.properties:
                 yield id_strings.search_property_locale(module, prop.name), trans(prop.label)
                 yield id_strings.search_property_hint_locale(module, prop.name), trans(prop.hint)
+                yield id_strings.search_property_validation_msg(module, prop.name), trans(prop.hint)
```
```suggestion
                yield id_strings.search_property_validation_msg(module, prop.name), trans(prop.validation[0].message)
```

esoergel replied:
:eyes:

orangejenny commented on corehq/apps/app_manager/id_strings.py:
```diff
@@ -274,6 +274,12 @@ def search_property_hint_locale(module, search_prop):
     return "search_property.m{module.id}.{search_prop}.hint".format(module=module, search_prop=search_prop)
 
 
+@pattern('search_property.m%d.%s.validation.message')
```
Could you incorporate the validation index here? It's a little confusing that the model supports multiple validation messages, but translations don't and the UI doesn't. I think it's fine for the UI not to support it, but translations feel to me like they should match the model.

esoergel replied:
good point, will do.

orangejenny commented on corehq/apps/app_manager/suite_xml/post_process/remote_requests.py:
```diff
@@ -274,6 +275,14 @@ def build_query_prompts(self):
+            if prop.validation:
+                kwargs['validation'] = [
+                    Validation(
+                        test=condition.xpath,
```
Calling `interpolate_xpath` on this will add support for `#user`, etc.

orangejenny replied:
Thinking about it, a `#search` or `#search_input` shortcut would be pretty nice. Though it's be a little different from `#user` and `#case`, I'd think we'd want to support something like `#search/email`, which would expand to `instance('search-input:results')/input/field[@name='email']`.

orangejenny reviewed (commented):
Minor comments. I'm ambivalent about the backend support for multiple validation conditions - my guess is that app builders will just use "and" (and lengthy validation messages) to combine conditions rather than spend the time in the future to build out a multi-condition UI.

esoergel committed:
Update corehq/apps/app_manager/app_strings.py

Co-authored-by: Jenny Schweers <orange.jenny@gmail.com>

esoergel committed:
Include validation index in id string

esoergel committed:
PR feedback

esoergel committed:
Codify means of storing translated labels

esoergel committed:
Configurable messages for required fields

I'm tempted to switch validation to be stored as a list of
`CustomAssertion`s, and for this to be a singleton of the same type,
though that'd require a (lazy) migration.

esoergel committed:
Validation condition xpaths should be added here

esoergel committed:
Make required_message updatable in the UI

esoergel committed:
Extract big block that calculates appearance

Might be nice to move this to the constructor of searchPropertyModel,
but that might make it less grokkable as a model definition

esoergel committed:
Use _.each instead of index vars

esoergel committed:
this fallback isn't needed

esoergel committed:
Use map and init list in one clean place

esoergel committed:
Use CustomAssertion for required and validations

esoergel committed:
Use consistent language in front-end

Though I do think text and test are similar enough to be confusing

esoergel committed:
Remove cruft leftover from the dev process

esoergel committed:
document assertion wackiness

At this point I honestly kinda wish I hadn't tried to reuse this code...

esoergel committed:
Merge pull request #31885 from dimagi/es/required

[USH] Case search: Custom messages for required conditions

esoergel commented:
I just merged in https://github.com/dimagi/commcare-hq/pull/31885
I'm treating this branch as already approved and will only be editing it by PR, so there shouldn't be a need to be any more reviewing here (though I'm happy to take feedback still if you've got it)

esoergel committed:
Remove validation when prop is hidden

https://dimagi-dev.atlassian.net/browse/QA-4384

esoergel committed:
Exclude null validation text from suite

https://dimagi-dev.atlassian.net/browse/QA-4378

esoergel committed:
Merge pull request #31915 from dimagi/es/validation-conditions-qa

Es/validation conditions qa
